[SDK-1914] Add a migration guide for v1 to v2 (#129)
auth0/express-openid-connect: V2_MIGRATION_GUIDE.md

> # V2 Migration Guide
> 

Reviewer: ```suggestion
# V2 Migration Guide

`v2.x` brings a number of breaking changes in the library behaviour, configuration options as well as its cookie format. As a result, `v1.x` session will not be accepted by the library after upgrading to `v2.x`, they will in fact be silently ignored and cleaned up.
```

## Configuration

> ### Required Keys

Reviewer: ```suggestion
### Required Configuration Properties
```

> - `appSession.secret` is now just `secret` (because it is now used to sign the transient cookies as well as the `appSession` cookie). The environmental variable has changed from `APP_SESSION_SECRET` to `SECRET`.

Reviewer: ```suggestion
- `appSession.secret` is now just `secret` (because it is now used to sign the transient cookies as well as the `appSession` cookie). The environment variable has changed from `APP_SESSION_SECRET` to `SECRET`.
```

```dotenv
# Before
SESSION_SECRET=LONG_RANDOM_STRING

# After
SECRET=LONG_RANDOM_STRING
```

```js
const { auth } = require('express-openid-connect');

// Before
app.use(
  auth({
    session: {
      secret: 'LONG_RANDOM_STRING',
    },
  })
);

// After
app.use(
  auth({
    secret: 'LONG_RANDOM_STRING',
  })
);
```

### Route customization

You can now enable individual routes rather than all or nothing and the path to routes can be customized using the `routes` config.

```js
// Before
app.use(
  auth({
    routes: true,
    loginPath: '/custom/login',
    logoutPath: '/custom/logout',
    redirectUriPath: '/custom/callback',
    postLogoutRedirectUri: '/custom/post-logout',
  })
);

// After
app.use(
  auth({
    routes: {
      login: '/custom/login',
      logout: '/custom/logout',
      callback: '/custom/callback',
      postLogoutRedirect: '/custom/post-logout',
    },
  })
);
```

### Session Lifecycle configuration

> Session duration was rolling only and configured using `appSession.duration` (default 24hrs). Now it can be configured to rolling or absolute (default rolling 24hrs and absolute 7 days).

Reviewer 2: I'm not familiar with rolling / absolute modes. Is this something that needs to be clarified here?

Reviewer 3: Should this be "sliding" instead of "rolling"? That's the more common terminology I believe.

Author: I thought we were happy with "rolling", but let me check to be safe

@panva @vibronet wdyt?

Reviewer: ```suggestion
Session duration was being refreshed (e.g. it was "rolling") for another 24 hours (default value unchanged) with every page visit and configured using `appSession.duration`. In addition to that being optional behaviour now (`appSession.rolling`) every session may also have a an absolute duration it will be "rolled" for, when that duration is passed the session is not accepted as valid anymore. The default for this "absolute" duration is 7 days.
```

```js
// Before
app.use(
  auth({
    appSession: {
      duration: 86400, // default 1 day in secs
    },
  })
);

// After
app.use(
  auth({
    appSession: {
      rolling: true,
      rollingDuration: 86400, // default 1 day rolling duration in secs
      absoluteDuration: 86400 * 7, // default 7 days absolute duration in secs
    },
  })
);
```

### `required` is now `authRequired`

To enable or disable all routes to require authentication, use the `authRequired` configuration (default `true`)

```js
// Before
app.use(
  auth({
    required: true,
  })
);

// After
app.use(
  auth({
    authRequired: true,
  })
);
```

### Configuration items renamed/removed

> - **`required` is now `authRequired`** - to enable or disable all routes to require authentication, use the `authRequired` configuration (default `true`)

Reviewer 2: this is explained right above this section. If it's OK to repeat it, wouldn't it be better to put it on the top of the guide, as a sort of high-level summary of the changes?

Author: I'll remove this line

- **`idTokenAlg` is now `idTokenSigningAlg`** - to specify an id token signing algorithm, use `idTokenSigningAlg`

> - **`httpOptions`** - it is no longer possible to pass custom http options to the underlying library

Reviewer: ```suggestion
- **`httpOptions`** was removed - it is no longer possible to pass custom http request options to the underlying library. It will be again in the future in a more curated and comprehensive feature we have in mind for this library.
```

> - **`handleCallback`** and **`getUser`** - The hooks have been removed and will be replaced by a collection of hooks in a later release.

Reviewer: ```suggestion
- **`handleCallback`** and **`getUser`** were removed - These "hooks" will be made available in the future in a more curated and comprehensive feature we have in mind for this library. ```

```js
// Before
app.use(
  auth({
    required: true,
    idTokenAlg: 'RS256',
    httpOptions: {},
    handleCallback: () => {},
    getUser: () => {},
  })
);

// After
app.use(
  auth({
    authRequired: true,
    idTokenSigningAlg: 'RS256',
  })
);
```

## Session and Context

This library adds an auth context to the request and response objects used within route handling called `oidc` (was `openid`).

```js
// Before
app.get('/', (req, res) => {
  const user = req.openid.user;
  const client = req.openid.client; // It is no longer possible to access the underlying client
  const isAuthenticated = req.isAuthenticated();
  const tokenSet = req.makeTokenSet({ tokens });

  // Login
  res.openid.login({});
  // Logout
  res.openid.logout({});
});

// After
app.get('/', async (req, res) => {
  const user = req.oidc.user;
  const claims = req.oidc.idTokenClaims;
  const isAuthenticated = req.oidc.isAuthenticated();
  const idToken = req.oidc.idToken;
  const {
    access_token,
    token_type,
    expires_in,
    isExpired,
    refresh,

>   } = req.oidc.accessToken; // If `code` in response_type

Reviewer 2: this is just an observation, but the property `oidc.accessToken` has an `access_token` property inside. Looks like it refers to the raw value of the token. In that case, if I want to access the access token value, it looks like I need to do `req.oidc.accessToken.access_token` 😕  shouldn't that be renamed to `raw` or `value`?

Also, why the `idToken` shares the same level with the `idTokenClaims` property, what's the reason to have it different from the access token?

Author: I'm happy with the API, but let me check to be safe

@panva @vibronet wdyt?

Reviewer: The reason why `accessToken` (named camelCase) is an object is that an Access Token is defined by more than just its value. It has a type, a possible expiration and optionally can be refreshed. That is why it's all "namespaced". I would be okay with renaming `accessToken.access_token` to `accessToken.value` if it makes sense to others. As is it has the structure of an OAuth Access Token Response - `{ access_token, expires_in, token_type }` with additional helper functions slapped in there.

I am indifferent to how `idToken` (holds the raw value) and `idTokenClaims` are exposed.

Author: ok - I'll check this with @vibronet and raise some additional work before the Beta launch if we want to change it

Reviewer 4: Interesting, good catch @lbalmaceda - `accessToken.access_token` does look odd. Renaming to `value` works for me, as long as we are consistent across the API surface (e.g. `idToken` is also a container, exposing `claims` and `value`)

  const refreshToken = req.oidc.refreshToken; // if `offline_access` in scope
  const userInfo = await req.oidc.fetchUserInfo();

  // Login
  res.oidc.login({});
  // Logout
  res.oidc.logout({});
});
```

## Custom Session Handling

> The ability to add custom session stores to the SDK using `appSession: false` has been removed and will be added back in a later release.

Reviewer 2: If we're confident this will eventually arrive, we can keep it. Otherwise, I suggest we use "could" to express that we might not add it at all.

Author: I'll change the wording

Reviewer: +1
